File: src/test/java/com/fwdekker/randomness/array/ArraySettingsTest.java
package com.fwdekker.randomness.array;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collections;

import static org.assertj.core.api.Assertions.assertThat;


/**
 * Unit tests for {@link ArraySettings}.
 */
final class ArraySettingsTest {
    private ArraySettings arraySettings;


    @BeforeEach
    void beforeEach() {
        arraySettings = new ArraySettings();
    }


    @Test
    void testGetLoadState() {
        arraySettings.setCount(997);
        arraySettings.setBrackets("0fWx<i6jTJ");
        arraySettings.setSeparator("f3hu)Rxiz1");
        arraySettings.setSpaceAfterSeparator(false);

        final ArraySettings newArraySettings = new ArraySettings();
        newArraySettings.loadState(arraySettings.getState());

        assertThat(newArraySettings.getCount()).isEqualTo(997);
        assertThat(newArraySettings.getBrackets()).isEqualTo("0fWx<i6jTJ");
        assertThat(newArraySettings.getSeparator()).isEqualTo("f3hu)Rxiz1");
        assertThat(newArraySettings.isSpaceAfterSeparator()).isEqualTo(false);
    }

    @Test
    void testGetSetCount() {
        arraySettings.setCount(655);

        assertThat(arraySettings.getCount()).isEqualTo(655);
    }

    @Test
    void testGetSetBrackets() {
        arraySettings.setBrackets("RLevljrzf0");

        assertThat(arraySettings.getBrackets()).isEqualTo("RLevljrzf0");
    }

    @Test
    void testGetSetSeparator() {
        arraySettings.setBrackets("d2[tlXkGf{");

        assertThat(arraySettings.getBrackets()).isEqualTo("d2[tlXkGf{");
    }

    @Test
    void testGetSetSpaceAfterSeparator() {
        arraySettings.setSpaceAfterSeparator(false);

        assertThat(arraySettings.isSpaceAfterSeparator()).isEqualTo(false);
    }


    @Test
    void testArrayifyEmpty() {
        assertThat(arraySettings.arrayify(Collections.emptyList()))
                .isEqualTo("[]");
    }

    @Test
    void testArrayify() {
        arraySettings.setCount(4);
        arraySettings.setBrackets("@#");
        arraySettings.setSeparator(";;");
        arraySettings.setSpaceAfterSeparator(true);

        assertThat(arraySettings.arrayify(Arrays.asList("Garhwali", "Pattypan", "Troll")))
                .isEqualTo("@Garhwali;; Pattypan;; Troll#");
    }

    @Test
    void testArrayifyNoBrackets() {
        arraySettings.setCount(8);
        arraySettings.setBrackets("");
        arraySettings.setSeparator("h");
        arraySettings.setSpaceAfterSeparator(false);

        assertThat(arraySettings.arrayify(Arrays.asList("Antheia", "Cowbinds", "Cotutor")))
                .isEqualTo("AntheiahCowbindshCotutor");
    }
}

File: src/test/kotlin/com/fwdekker/randomness/array/ArraySettingsTest.kt
package com.fwdekker.randomness.array

import org.assertj.core.api.Assertions.assertThat
import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.specification.describe


/**
 * Unit tests for [ArraySchemeDecorator].
 */
object ArraySettingsTest : Spek({
    lateinit var arraySchemeDecorator: ArraySchemeDecorator


    beforeEachTest {
        arraySchemeDecorator = ArraySchemeDecorator()
    }


    describe("copying") {
        it("creates an independent copy") {
            val copy = arraySchemeDecorator.deepCopy()
            arraySchemeDecorator.currentScheme.count = 44
            copy.currentScheme.count = 15

            assertThat(arraySchemeDecorator.currentScheme.count).isEqualTo(44)
        }

        it("copies state from another instance") {
            arraySchemeDecorator.currentScheme.count = 997
            arraySchemeDecorator.currentScheme.brackets = "0fWx<i6jTJ"
            arraySchemeDecorator.currentScheme.separator = "f3hu)Rxiz1"
            arraySchemeDecorator.currentScheme.isSpaceAfterSeparator = false

            val newArraySchemeDecorator = ArraySchemeDecorator()
            newArraySchemeDecorator.copyFrom(arraySchemeDecorator.state)

            assertThat(newArraySchemeDecorator.currentScheme.count).isEqualTo(997)
            assertThat(newArraySchemeDecorator.currentScheme.brackets).isEqualTo("0fWx<i6jTJ")
            assertThat(newArraySchemeDecorator.currentScheme.separator).isEqualTo("f3hu)Rxiz1")
            assertThat(newArraySchemeDecorator.currentScheme.isSpaceAfterSeparator).isEqualTo(false)
        }
    }
})


/**
 * Unit tests for [ArraySchemeDecorator].
 */
object ArraySchemeTest : Spek({
    lateinit var arraySchemeDecorator: ArraySchemeDecorator


    beforeEachTest {
        arraySchemeDecorator = ArraySchemeDecorator()
    }


    describe("arrayify") {
        it("returns only the brackets if the array is empty") {
            assertThat(arraySchemeDecorator.arrayify(emptyList())).isEqualTo("[]")
        }

        it("arrayifies an array") {
            arraySchemeDecorator.count = 4
            arraySchemeDecorator.brackets = "@#"
            arraySchemeDecorator.separator = ";;"
            arraySchemeDecorator.isSpaceAfterSeparator = true

            assertThat(arraySchemeDecorator.arrayify(listOf("Garhwali", "Pattypan", "Troll")))
                .isEqualTo("@Garhwali;; Pattypan;; Troll#")
        }

        it("arrayifies an array without brackets") {
            arraySchemeDecorator.count = 8
            arraySchemeDecorator.brackets = ""
            arraySchemeDecorator.separator = "h"
            arraySchemeDecorator.isSpaceAfterSeparator = false

            assertThat(arraySchemeDecorator.arrayify(listOf("Elvish", "Stride", "Bills"))).isEqualTo("ElvishhStridehBills")
        }

        it("does not place a space after separator if that option is false") {
            arraySchemeDecorator.count = 3
            arraySchemeDecorator.brackets = "<>"
            arraySchemeDecorator.separator = "-"
            arraySchemeDecorator.isSpaceAfterSeparator = false

            assertThat(arraySchemeDecorator.arrayify(listOf("Remain", "Pound"))).isEqualTo("<Remain-Pound>")
        }

        it("ignores space after separator if newline separator is used") {
            arraySchemeDecorator.count = 2
            arraySchemeDecorator.brackets = "[]"
            arraySchemeDecorator.separator = "\n"
            arraySchemeDecorator.isSpaceAfterSeparator = true

            assertThat(arraySchemeDecorator.arrayify(listOf("Union", "Bell"))).isEqualTo("[Union\nBell]")
        }
    }

    describe("copying") {
        describe("copyFrom") {
            it("makes the two schemes equal") {
                val schemeA = ArraySchemeDecorator()
                val schemeB = ArraySchemeDecorator(myName = "Name")
                assertThat(schemeA).isNotEqualTo(schemeB)

                schemeA.copyFrom(schemeB)

                assertThat(schemeA).isEqualTo(schemeB)
            }
        }

        describe("copyAs") {
            it("makes two schemes equal except for the name") {
                val schemeA = ArraySchemeDecorator()
                val schemeB = schemeA.copyAs("NewName")
                assertThat(schemeA).isNotEqualTo(schemeB)

                schemeB.myName = schemeA.myName

                assertThat(schemeA).isEqualTo(schemeB)
            }
        }
    }
})
